### PR TITLE
security: harden API, webhook, CI, and agent instructions

### PR DESCRIPTION
**Pull request by robert-cronin**
## Description

Security hardening across the backend API, controller webhook, CI pipelines, frontend, provider transformers, and agent instructions. Addresses findings from a comprehensive security audit covering injection, authz, supply chain, and AI/agentic attack vectors.

## AI Prompt (Optional)

<details>
<summary>🤖 AI Prompt Used</summary>

```
I want to strengthen the security of this whole project. Spin up subagents to audit the security posture of the entire project, especially given everyone will use agentic programming a lot to code, so focus specifically on AI attack vectors and audit the agents.md and claude.md too.
```

Then after the audit report:

```
Fix the findings and I'll push it as a PR.
```

**AI Tool:** GitHub Copilot CLI (Claude Opus 4.6)

</details>

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [x] 📚 Documentation update
- [ ] 🎨 UI/UX improvement
- [ ] ♻️ Refactoring (no functional changes)
- [x] 🧪 Test update
- [x] 🔧 Build/CI configuration

## Related Issues

Relates to overall project security posture — no specific issue.

### Backend API (`backend/src/`)
- **CORS restricted**: Changed default from `origin: '*'` to `http://localhost:5173` (configurable via `CORS_ORIGIN` env var, supports comma-separated list)
- **Error leakage stopped**: Unhandled errors now return generic `"Internal Server Error"` instead of leaking `err.message` with internal details
- **OAuth verifier secured**: Deprecated the public `GET /verifier/:state` endpoint (returns 410). Added new `POST /token-with-state` endpoint that keeps the PKCE verifier server-side — the verifier never leaves the server
- **Health endpoints restricted**: `/api/health/nodes` and `/api/health/status` now require auth when `AUTH_ENABLED=true`. Only `/api/health`, `/api/health/version`, and `/api/cluster/status` remain public

### Controller Webhook (`controller/internal/webhook/`)
- **Override allowlist**: New `validateOverrides()` function blocks dangerous fields in `spec.provider.overrides` — `securityContext`, `hostNetwork`, `hostPID`, `hostIPC`, `serviceAccountName`, `serviceAccount`, `automountServiceAccountToken`, `nodeName`, `priorityClassName`, `runtimeClassName`. Works recursively on nested objects
- **Resource ceilings**: New `validateResourceQuantity()` enforces maximum bounds — GPU count (64), replicas (32), CPU (512 cores), memory (4Ti)

### Provider Security (`providers/`)
- **Pod security context defaults**: Added `allowPrivilegeEscalation: false` and `capabilities.drop: ["ALL"]` to containers created by kaito (llamacpp template) and llmd (vllm containers)

### CI/CD (`.github/workflows/`)
- **Explicit permissions**: Added `permissions: contents: read` to `test.yml`, `e2e-backend.yml`, `e2e-controller.yml`, `e2e-gateway.yml`

### Frontend (`frontend/`)
- **Content Security Policy**: Added a production-only Vite build plugin that injects a CSP meta tag into the built `index.html` — restricts `default-src`, `script-src`, `frame-src`, `object-src` to `'self'`; allows fonts from existing CDNs; restricts `connect-src` to `'self'` plus the exact `VITE_API_URL` origin when one is configured

### Agent Instructions
- **Security rules section**: Added mandatory security rules to `agents.md` — never push/merge without approval, never deploy to production, never leak secrets, treat external input as untrusted, sanitize user input
- **Deploy skill safety gate**: Added cluster context verification step to `.agents/skills/deploy-controller/SKILL.md` — requires explicit human confirmation before any deploy operation

## Testing

- [x] Unit tests pass (`bun run test`)
- [x] Manual testing performed
- [x] Tested with a Kubernetes cluster

### Automated tests
- **15 new webhook unit tests** (`security_validation_test.go`): Override blocklist (securityContext, hostNetwork, serviceAccountName, nested keys, safe fields, nil, invalid JSON), resource quantity validation (within/exceeds CPU, memory, empty, invalid format), GPU ceiling enforcement
- **683 backend tests pass** (0 fail) — including updated OAuth flow tests for the deprecated `/verifier/:state` (410) and new `/token-with-state` endpoint, plus CORS coverage for `localhost`, `127.0.0.1`, and `[::1]` loopback origins
- **117 frontend tests pass**
- **Frontend production build passes** — verified default CSP emits `connect-src 'self'` and `VITE_API_URL=https://api.example.test/v1` emits `connect-src 'self' https://api.example.test`
- **Backend build passes**
- **All 4 provider test suites pass** (kaito, dynamo, kuberay, llmd)

### Manual cluster validation (Kind)
Created a Kind cluster, built the controller image, deployed with webhooks enabled, and ran live admission tests:

| Test | Expected | Result |
|------|----------|--------|
| `spec.provider.overrides.securityContext: {privileged: true}` | REJECTED | ✅ `Forbidden: overriding "securityContext" is not allowed` |
| `spec.provider.overrides.hostNetwork: true` | REJECTED | ✅ `Forbidden: overriding "hostNetwork" is not allowed` |
| `spec.provider.overrides.replicas: 2` (safe field) | ACCEPTED | ✅ Created successfully |
| `spec.resources.gpu.count: 65` | REJECTED | ✅ `GPU count exceeds maximum allowed (64)` |
| `spec.resources.gpu.count: 8` | ACCEPTED | ✅ Created successfully |
| Nested `spec.provider.overrides.spec.serviceAccountName: cluster-admin` | REJECTED | ✅ `Forbidden: overriding "serviceAccountName" is not allowed` |

## Checklist

- [x] My code follows the project's style guidelines
- [ ] I have run `bun run lint`
- [x] I have added tests that prove my fix/feature works
- [x] New and existing unit tests pass locally
- [x] I have updated documentation if needed
- [x] My changes generate no new warnings

## Screenshots

N/A — security hardening, no UI changes.

### What this PR does NOT fix (requires larger architectural changes)
These findings from the audit are noted but out of scope for this PR:
- **localStorage → cookie auth migration** — breaking change to auth flow
- **Full role-based authorization** — major backend feature (viewer vs admin)
- **Dockerfile digest pinning** — operational process, needs CI integration
- **CEL expression caching/complexity limits** — controller internals, needs careful testing
- **Gateway namespace auto-expansion restriction** — could break existing deployments
- **VEX workflow lock file fixes** — auto-generated file
- **Rate limiting** — needs middleware selection and configuration design
- **Secret ref validation in providers** — needs policy design for cross-namespace secret access

### Breaking change note
The `GET /api/oauth/huggingface/verifier/:state` endpoint now returns 410. The Headlamp plugin or any client using this endpoint should migrate to `POST /api/oauth/huggingface/token-with-state` which accepts `{code, state}` and handles the verifier server-side. The original `POST /api/oauth/huggingface/token` endpoint (with client-supplied `codeVerifier`) still works for backward compatibility.

### ESLint
`bun run lint` was attempted, but it currently fails before checking the changed code because the repo is missing a frontend ESLint configuration under `frontend/src` (pre-existing repo/config issue).

### CODE REVIEW

**Comment by surajssd** [comment]
@robert-cronin can you rebase?

File: frontend/index.html [hunk 1]
```diff
@@ -5,6 +5,7 @@
     <link rel="icon" type="image/x-icon" href="/favicon.ico" />
     <link rel="icon" type="image/png" sizes="32x32" href="/favicon.png" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta http-equiv="Content-Security-Policy" content="default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline' https://api.fontshare.com https://fonts.googleapis.com; font-src 'self' https://api.fontshare.com https://fonts.gstatic.com; img-src 'self' data: https:; connect-src 'self'; frame-src 'none'; object-src 'none'; base-uri 'self';" />
```
[the remark this answers is not in the document]
Comment (sozercan) [reply frontend/index.html]: Addressed in 87eb4dc: CSP is now injected only during production Vite builds via transformIndexHtml, so dev/HMR runs without the strict policy, and production connect-src allows http:/https: for configured API origins.

File: controller/internal/webhook/v1alpha1/modeldeployment_webhook.go [hunk 2]
```diff
@@ -231,6 +244,40 @@ func (v *ModelDeploymentCustomValidator) validateSpec(obj *airunwayv1alpha1.Mode
...
+		}
+		if spec.Scaling.Decode != nil {
+			if spec.Scaling.Decode.Replicas > MaxReplicas {
+				allErrs = append(allErrs, field.Invalid(specPath.Child("scaling", "decode", "replicas"), spec.Scaling.Decode.Replicas, fmt.Sprintf("exceeds maximum replicas (%d)", MaxReplicas)))
+			}
+			if spec.Scaling.Decode.GPU != nil && spec.Scaling.Decode.GPU.Count > MaxGPUCount {
+				allErrs = append(allErrs, field.Invalid(specPath.Child("scaling", "decode", "gpu", "count"), spec.Scaling.Decode.GPU.Count, fmt.Sprintf("exceeds maximum GPU count (%d)", MaxGPUCount)))
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply controller/internal/webhook/v1alpha1/modeldeployment_webhook.go:281]: Addressed in 87eb4dc: added a spec.scaling.replicas MaxReplicas check and coverage.

File: controller/internal/webhook/v1alpha1/modeldeployment_webhook.go [hunk 3]
```diff
@@ -231,6 +244,40 @@ func (v *ModelDeploymentCustomValidator) validateSpec(obj *airunwayv1alpha1.Mode
...
+			}
+		}
+		if spec.Scaling.Decode != nil {
+			if spec.Scaling.Decode.Replicas > MaxReplicas {
+				allErrs = append(allErrs, field.Invalid(specPath.Child("scaling", "decode", "replicas"), spec.Scaling.Decode.Replicas, fmt.Sprintf("exceeds maximum replicas (%d)", MaxReplicas)))
+			}
+			if spec.Scaling.Decode.GPU != nil && spec.Scaling.Decode.GPU.Count > MaxGPUCount {
+				allErrs = append(allErrs, field.Invalid(specPath.Child("scaling", "decode", "gpu", "count"), spec.Scaling.Decode.GPU.Count, fmt.Sprintf("exceeds maximum GPU count (%d)", MaxGPUCount)))
+			}
+		}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply controller/internal/webhook/v1alpha1/modeldeployment_webhook.go:280]: Addressed in 87eb4dc: added MaxMemory validation for scaling.prefill.memory and scaling.decode.memory plus tests.

File: backend/src/hono-app.ts [hunk 4]
```diff
@@ -192,10 +223,11 @@ app.onError((err, c) => {
     );
   }
 
+  // Don't leak internal error details to clients
   return c.json(
     {
       error: {
-        message: err.message || 'Internal Server Error',
+        message: 'Internal Server Error',
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.ts:264]: Addressed in 87eb4dc: the global error handler now sanitizes 5xx HTTPException responses while still preserving non-5xx messages.

File: backend/src/hono-app.ts [hunk 5]
```diff
@@ -54,7 +70,7 @@ app.use('*', compress());
 app.use(
   '*',
   cors({
-    origin: CORS_ORIGIN,
+    origin: parseCorsOrigin(CORS_ORIGIN),
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.ts]: Addressed in 87eb4dc: default CORS now accepts loopback browser origins (localhost, 127.0.0.1, [::1]) for Vite/Headlamp local clients without reopening wildcard CORS; CORS_ORIGIN remains configurable for deployed hosts and is documented.

File: frontend/vite.config.ts [hunk 6]
```diff
@@ -1,10 +1,40 @@
...
+  "base-uri 'self'",
+].join('; ')
+
+function cspPlugin(): Plugin {
+  return {
+    name: 'inject-csp-meta',
+    apply: 'build',
+    transformIndexHtml(html) {
+      const tag = `<meta http-equiv="Content-Security-Policy" content="${PROD_CSP}" />`
+      return html.replace('<meta name="viewport"', `${tag}\n    <meta name="viewport"`)
```
[the remark this answers is not in the document]
Comment (sozercan) [reply frontend/vite.config.ts]: Addressed in 87eb4dc: switched CSP injection to Vite structured transformIndexHtml tag return with injectTo: head-prepend.

File: backend/src/hono-app.ts [hunk 7]
```diff
@@ -83,9 +123,16 @@ app.use('/api/*', async (c, next) => {
...
+  // Skip auth for exact-match public routes
   const path = c.req.path;
-  if (PUBLIC_ROUTES.some(route => path === route || path.startsWith(route + '/'))) {
+  if (PUBLIC_ROUTES_EXACT.includes(path)) {
+    return next();
+  }
+
+  // Skip auth for prefix-match public routes (cluster/status, settings, oauth)
+  if (PUBLIC_ROUTES.some(route =>
+    route !== '/api/health' && (path === route || path.startsWith(route + '/'))
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.ts]: Addressed in 87eb4dc: removed /api/health from the prefix allowlist and rely on the exact-match public health routes.

File: controller/internal/webhook/v1alpha1/modeldeployment_webhook.go [hunk 8]
```diff
@@ -692,3 +744,97 @@ func (v *ModelDeploymentCustomValidator) validateStorage(obj *airunwayv1alpha1.M
...
+
+	var overrideMap map[string]interface{}
+	if err := json.Unmarshal(spec.Provider.Overrides.Raw, &overrideMap); err != nil {
+		// Don't echo the raw payload back: it can be large and may contain
+		// data the user didn't expect to see in admission errors/logs.
+		allErrs = append(allErrs, field.Invalid(
+			specPath.Child("provider", "overrides"),
+			fmt.Sprintf("<redacted %d bytes>", len(spec.Provider.Overrides.Raw)),
+			"overrides must be valid JSON",
+		))
```
[the remark this answers is not in the document]
Comment (sozercan) [reply controller/internal/webhook/v1alpha1/modeldeployment_webhook.go]: Addressed in 87eb4dc: overrides now unmarshal into interface{} and explicitly require a JSON object, with tests for array/string/number/null payloads.

File: backend/src/hono-app.ts [hunk 9]
```diff
@@ -45,7 +45,58 @@ logger.info(
 // Main App
 // ============================================================================
 
-const CORS_ORIGIN = process.env.CORS_ORIGIN || '*';
+const DEFAULT_CORS_ORIGINS = ['http://localhost:5173'];
+const DEFAULT_CORS_ORIGIN = DEFAULT_CORS_ORIGINS.join(',');
+const LOOPBACK_HOSTS = new Set(['localhost', '127.0.0.1', '[::1]']);
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.ts]: Addressed in 775d5d6: `LOOPBACK_HOSTS` now stores bare `::1`, and `isLoopbackOrigin()` normalizes any bracketed IPv6 hostname before checking the loopback allowlist. I also added coverage for `http://[::1]:4466` alongside `localhost` and `127.0.0.1`.

File: backend/src/hono-app.test.ts [hunk 10]
```diff
@@ -23,6 +32,25 @@ describe('Hono Routes', () => {
...
+      const loopbackOrigin = 'http://localhost:4466';
+      const loopbackRes = await app.request('/api/health', {
+        headers: { Origin: loopbackOrigin },
+      });
+      expect(loopbackRes.headers.get('Access-Control-Allow-Origin')).toBe(loopbackOrigin);
+
+      const externalRes = await app.request('/api/health', {
+        headers: { Origin: 'https://example.com' },
+      });
+      expect(externalRes.headers.get('Access-Control-Allow-Origin')).toBeNull();
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.test.ts]: Addressed in 775d5d6: the CORS test no longer silently returns when `CORS_ORIGIN` is set. It now uses an explicit `test.skip` path for that environment-dependent case, so CI output makes the skip visible instead of masking a no-op.

File: frontend/vite.config.ts [hunk 11]
```diff
@@ -1,10 +1,48 @@
...
 import react from '@vitejs/plugin-react'
 import path from 'path'
 
+// Inject a strict Content-Security-Policy meta tag into index.html for
+// production builds only. In `vite dev`, React refresh injects inline scripts
+// and HMR uses ws:// connections, both of which a strict CSP would block.
+// Keeping the CSP build-only lets dev work normally while production gets the
+// strict policy. connect-src is intentionally permissive (http:/https:) so
+// deployments that point VITE_API_URL at a different origin still work; tighten
+// further per-deployment via reverse-proxy headers if needed.
```
[the remark this answers is not in the document]
Comment (sozercan) [reply frontend/vite.config.ts]: Addressed: I updated the PR description to match the implementation. It now describes the production-only Vite build plugin that injects the CSP meta tag into the built `index.html`, rather than saying the tag was added directly to `frontend/index.html`.

File: frontend/vite.config.ts [hunk 12]
```diff
@@ -1,10 +1,48 @@
...
+const PROD_CSP = [
+  "default-src 'self'",
+  "script-src 'self'",
+  "style-src 'self' 'unsafe-inline' https://api.fontshare.com https://fonts.googleapis.com",
+  "font-src 'self' https://api.fontshare.com https://cdn.fontshare.com https://fonts.gstatic.com",
+  "img-src 'self' data: https:",
+  "connect-src 'self' http: https:",
```
[the remark this answers is not in the document]
Comment (sozercan) [reply frontend/vite.config.ts]: Addressed in 775d5d6: production `connect-src` is now restricted to `'self'` plus the exact origin derived from a valid absolute `http(s)` `VITE_API_URL` when configured. Relative API paths remain covered by `'self'`, and invalid/unsupported API URLs are ignored with a build warning instead of widening the policy.

File: backend/src/hono-app.ts [hunk 13]
```diff
@@ -45,7 +45,61 @@ logger.info(
...
+    .map((o) => o.trim())
+    .filter((o) => o.length > 0);
+  if (list.length > 0) return list;
+  // Fail closed: a malformed CORS_ORIGIN (e.g. ",,") should keep the secure
+  // default rather than broaden access to '*'.
+  logger.warn(
+    { rawCorsOrigin: raw },
+    `CORS_ORIGIN is set but parses to no origins; falling back to ${DEFAULT_CORS_ORIGIN}`,
+  );
+  return DEFAULT_CORS_ORIGINS;
```
[the remark this answers is not in the document]
Comment (sozercan) [reply backend/src/hono-app.ts]: Addressed in 169fbfa: `parseCorsOrigin()` now falls back to the same `defaultCorsOrigin` loopback function when `CORS_ORIGIN` is set but parses to no origins. That means empty/malformed values now allow localhost/127.0.0.1/[::1] on any port and continue to reject external origins. I also added a regression test covering the empty `CORS_ORIGIN` fallback.

File: deploy/README.md [hunk 14]
```diff
@@ -110,6 +111,22 @@ env:
...
+allowed origin or a comma-separated allowlist:
+
+```yaml
+env:
+  - name: CORS_ORIGIN
+    value: "https://airunway.example.com,https://admin.example.com"
+```
+
+Use `CORS_ORIGIN="*"` only when you intentionally want to allow every browser
+origin. Empty or malformed values fall back to the safe loopback-only default.
```
[the remark this answers is not in the document]
Comment (sozercan) [reply deploy/README.md:128]: Addressed in 169fbfa by aligning the implementation with the docs: empty/malformed `CORS_ORIGIN` now falls back to the broader loopback-only default (`localhost`, `127.0.0.1`, and `[::1]` on any port) instead of only `http://localhost:5173`, so the deploy README wording now matches the backend behavior.

**Review by sozercan — APPROVED** [review]
